File: akka_essentials/src/main/java/chapter2java8/actor/AggregateActor.java
package chapter2java8.actor;

import akka.actor.AbstractActor;
import akka.japi.pf.ReceiveBuilder;
import chapter2.message.ReduceData;
import chapter2.message.Result;

import java.util.HashMap;
import java.util.Map;

import static akka.japi.pf.ReceiveBuilder.match;

public class AggregateActor extends AbstractActor {

    private Map<String, Integer> finalReducedMap = new HashMap<>();

    public AggregateActor() {
        receive(ReceiveBuilder.
                match(ReduceData.class, reduceData -> aggregateInMemoryReduce(reduceData.getReduceDataList()))
                .match(Result.class, result -> sender().tell(finalReducedMap.toString(), self()))
                .build());
    }

    private void aggregateInMemoryReduce(Map<String, Integer> reducedList) {
        for (String key : reducedList.keySet()) {
            Integer count = reducedList.get(key) + finalReducedMap.getOrDefault(key, 0);
            finalReducedMap.put(key, count);
        }
    }

}

File: akka_essentials/src/main/scala/essentials/actor/scala/chapter2/actor/AggregateActor.scala
package essentials.actor.scala.chapter2.actor

import akka.actor.Actor
import essentials.actor.scala.chapter2.{ReduceData, Result}

import scala.collection.mutable

class AggregateActor extends Actor {

  val finalReducedMap = new mutable.HashMap[String, Int]

  def receive: Receive = {
    case ReduceData(reduceDataMap) => aggregateInMemoryReduce(reduceDataMap)
    case Result => sender ! finalReducedMap.toString()
  }

  def aggregateInMemoryReduce(reducedList: Map[String, Int]): Unit = {
    for ((key, value) <- reducedList) {
      finalReducedMap += (key -> (value + finalReducedMap.getOrElse(key, 0)))
    }
  }

}
